runtime/compiler/optimizer/abstractinterpreter/AbsInterpreter.cpp

```diff
@@ -0,0 +1,3072 @@
+         value->setParameterPosition(paramPos++);
+         state->set(slotIndex++, value);
+         }
+
+      //set the explicit parameters
+      for (TR_MethodParameterIterator *pi = _callerMethod->getParameterIterator(*comp()); !pi->atEnd(); pi->advanceCursor(), slotIndex++, paramPos++)
+         {
+         TR::DataType dataType = pi->getDataType();
+         TR::AbsValue* param = NULL;
+
+         switch (dataType)
+            {
+            case TR::Int8:
+            case TR::Int16:
+            case TR::Int32:
+               param = createTopInt();
+               param->setParameterPosition(paramPos);
+               state->set(slotIndex, param);
+               break;
+
+            case TR::Int64:
+               param = createTopLong();
+               param->setParameterPosition(paramPos);
+               state->set(slotIndex, param);
+               slotIndex++;
+               state->set(slotIndex, createTopLong());
+               break;
+
+            case TR::Double:
+               param = createTopDouble();
+               param->setParameterPosition(paramPos);
+               state->set(slotIndex, param);
+               slotIndex++;
+               state->set(slotIndex, createTopDouble());
+               break;
+
+            case TR::Float:
+               param = createTopFloat();
+               param->setParameterPosition(paramPos);
+               state->set(slotIndex, param);
+               break;
+
+            case TR::Aggregate:
+               {
+               TR_OpaqueClassBlock *classBlock = pi->getOpaqueClass();
+               if (pi->isArray())
+                  {
+                  int32_t arrayType = comp()->fe()->getNewArrayTypeFromClass(classBlock);
+                  int32_t elemetSize = arrayType == 7 || arrayType == 11 ? 8 : 4; //7: double, 11: long
+                  param = createArrayObject(classBlock, TR_maybe, 0, INT32_MAX, elemetSize);
+                  param->setParameterPosition(paramPos);
+                  state->set(slotIndex, param);
+                  }
+               else
+                  {
+                  param = createObject(classBlock, TR_maybe);
+                  param->setParameterPosition(paramPos);
+                  state->set(slotIndex, param);
+                  }
+               break;
+               }
+
+            default:
+               TR_ASSERT_FATAL(false, "Uncatched type");
+               break;
+            }
+         }
+      }
+
+   if (comp()->getOption(TR_TraceAbstractInterpretation))
+      {
+      traceMsg(comp(), "Start Block state of method %s:\n", _callerMethod->signature(comp()->trMemory()));
+      state->print(comp());
+      }
+
+   return state;
+   }
+
+bool TR::AbsBlockInterpreter::interpret()
+   {
+   if (!getState()) // the block is in a dead code area. No need to interpret.
+      return true;
+
+   while (_bci.currentByteCodeIndex() < getBlockEndIndex() && _bci.current() != J9BCunknown)
```

Comment:
`TR::AbsInterpreter::interpret()` firstly run `TR::ReversePostorderSnapshotBlockIterator` to generate a block iterator. Then, it will traverse the block iterater and for each block, it will call `TR::AbsInterpreter::interpretBlock()`. This then calls the constructor `TR::AbsBlockInterpreter::AbsBlockInterpreter()` to create a block interpreter. When initiallizing the block interpreter, it set `_bci` to the start index of the block by `_bci.setIndex(getBlockStartIndex());`, where `int32_t getBlockStartIndex() { return _block->getBlockBCIndex(); }`. Finally, `TR::AbsInterpreter::interpretBlock()` calls `TR::AbsBlockInterpreter::interpret()` to interprete each bytecode in the block.